### PR TITLE
Update distribution spec for inner child of left outer index NL join to handle random distributions

### PR DESCRIPTION
**Pull request by danhoffman699**
Previously, CPhysicalLeftOuterIndexNLJoin requested any distribution on the
inner child, and threw an exception (and fell back to planner) if the inner
table was randomly distributed to prevent incorrect results.  This commit adds
CPhysicalLeftOuterIndexNLJoin::EpetDistribution to prohibit certain plans so we
do not reach an invalid state (i.e. reject random distributions on the inner
child without creating a new distribution spec entirely).

This commit adds a specific case where `EpetDistribution` is called if we are
using `CPhysicalLeftOuterIndexNLJoin` (normally, an operation with
`CDistributionSpecAny` will not have `EpetDistribution` called, so we made an
exception specifc to this operation) and adds test cases to `gporca` to verify the
behavior is correct. This commit also adds missing test cases for hashed and
replicated tables with left outer index nested loop joins.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/test/regress/expected/gporca.out [hunk 1]
```diff
@@ -13692,3 +13692,160 @@ select data from tt_varchar where data > any(select id from tt_int);
...
+select r.a, r.b, r.c, l.c from left_outer_index_nl_foo_repl r left outer join left_outer_index_nl_bar_repl l on r.b=l.b;
+ a | b | c | c 
+---+---+---+---
+ 1 | 1 | 1 |  
+ 2 | 2 | 2 | 2
+ 3 | 3 | 3 | 3
+ 4 | 4 | 4 | 4
+(4 rows)
+
+--- outer side replicated, inner side hashed can have interesting cases
```
Comment (bhuvnesh2703) [inline src/test/regress/expected/gporca.out:13823]: could you elaborate here what you refer by interesting cases?

**Review by bhuvnesh2703 — APPROVED** [review]
LGTM, other than the minor comment to explain the interesting cases.

File: src/test/regress/expected/gporca.out (hunk 1 above)
Comment (danhoffman699) [reply src/test/regress/expected/gporca.out]: Chris mentioned this off hand, but I added what I think he meant

File: src/backend/gporca/libgpopt/src/operators/CPhysicalLeftOuterIndexNLJoin.cpp [hunk 2]
```diff
@@ -52,6 +52,29 @@ CPhysicalLeftOuterIndexNLJoin::Matches(COperator *pop) const
 	return false;
 }
 
+CEnfdProp::EPropEnforcingType
+CPhysicalLeftOuterIndexNLJoin::EpetDistribution(
+	CExpressionHandle &exprhdl,			 // exprhdl
+	const CEnfdDistribution *ped) const	 // ped
```
Comment (dgkimura) [inline src/backend/gporca/libgpopt/src/operators/CPhysicalLeftOuterIndexNLJoin.cpp:58]: Nit: These commented variables don't seem useful.

File: src/test/regress/sql/gporca.sql [hunk 3]
```diff
@@ -3268,3 +3268,62 @@ reset optimizer_enforce_subplans;
 -- start_ignore
 DROP SCHEMA orca CASCADE;
 -- end_ignore
+
+create table left_outer_index_nl_foo (a integer, b integer, c integer) distributed randomly;
```
Comment (dgkimura) [inline src/test/regress/sql/gporca.sql:3272]: Nit: Seems like this test intentionally created orca schema at beginning of test file and drops at the end.  Any reason not to follow that pattern?

**Review by dgkimura — COMMENTED** [review]
Looks good to me.  Minor comments.

Also, could you briefly conceptually explain why left outer index nested loop join with random inner child produces wrong results and why that isn't the case for non-index scenario?

File: src/backend/gporca/libgpopt/src/engine/CEngine.cpp [hunk 4]
```diff
@@ -2092,9 +2092,21 @@ CEngine::FCheckEnfdProps(CMemoryPool *mp, CGroupExpression *pgexpr,
...
 					  !prpp->Peo()->PosRequired()->IsEmpty();
 
+	// CPhysicalLeftOuterIndexNLJoin requires the inner child to be any
+	// distribution but random. The OR makes an exception in this case.
+	// This should be generalized when more physical operators require
+	// this pattern. We need an explicit check for CPhysicalLeftOuterIndexNLJoin
+	// when there are no motions, therefore we need to handle this exceptional
+	// case here.
+	//
+	// Similar exceptions should be OR'd into fDistirbutionReqdException to
```
[the remark this answers is not in the document]
Comment (danhoffman699) [reply src/backend/gporca/libgpopt/src/engine/CEngine.cpp:2102]: fixed in the latest commit

File: src/backend/gporca/libgpopt/src/engine/CEngine.cpp [hunk 5]
```diff
@@ -2092,9 +2092,21 @@ CEngine::FCheckEnfdProps(CMemoryPool *mp, CGroupExpression *pgexpr,
...
+	// force checking EpetDistribution on the physical operation
+	BOOL fDistributionReqdException =
+		popPhysical->Eopid() == COperator::EopPhysicalLeftOuterIndexNLJoin;
 	BOOL fDistributionReqd =
-		!GPOS_FTRACE(EopttraceDisableMotions) &&
-		(CDistributionSpec::EdtAny != prpp->Ped()->PdsRequired()->Edt());
+		(!GPOS_FTRACE(EopttraceDisableMotions) &&
+		 (CDistributionSpec::EdtAny != prpp->Ped()->PdsRequired()->Edt())) ||
+		fDistributionReqdException;
```
[the remark this answers is not in the document]
Comment (danhoffman699) [reply src/backend/gporca/libgpopt/src/engine/CEngine.cpp:2109]: This was intentional at the time, but is bad design in hindsight. Fixed in the latest commit

**Pull request merged**